### PR TITLE
Finite Element Exterior Calculus

### PR DESCRIPTION
**Pull request by yguclu**
- Add discrete DeRham sequence operators
- Add 1D/2D/3D push-forwards and pull-backs
- Add 1D/2D/3D global projectors
- Remove Analytical mapping from library, use the one from Sympde instead
- Modify the `VectorFemField` to be callable
- Fix bug in the assembly of periodic basis functions
- Add `KroneckerStencilMatrix` 
- Add Identity operators
- Add tests when pads are different from the self.domain.pads
- Add `transpose()` method to `KroneckerStencilMatrix` and `BlockMatrix`
- Fix `transpose()` method of `StencilMatrix` in the case where the pads are different from self.domain.pads
- Add `tostencil()` method to `KroneckerStencilMatrix`
- Add conversion to PETSc data structure
- Add 1D FEEC unit test w/ periodic and Dirichlet BCs (can also be run as a script)
- Add 3D FEEC unit test w/ periodic BCs

TODO (future work):
- Implement systematic strategy for FEEC with homogeneous Dirichlet BCs on V0 space
- Add 2D FEEC unit test w/ periodic and Dirichlet BCs
- Add 3D FEEC unit test w/ Dirichlet BCs
- Verify that everything works in parallel

### CODE REVIEW

**Comment by yguclu** [comment]
> Ok, so I guess that at some point the 'normalization' argument will be called 'basis' ?

I wanted to call it `basis`, but in this low-level module the name `basis` is used to identify an array variable that contains the values of the basis functions. So for now I decided to call it `normalization` instead, in order to avoid confusion. But as you see it is not a boolean variable any more, instead it is a string which can take the possible values `'B'` or `'M'`.

In the `SplineSpace` class in **psydac/fem/splines.py**, which is higher level, I pass the argument `basis` to the constructor. Again, this takes one of the possible values `'B'` or `'M'`.

**Comment by yguclu** [comment]
> A spline space of degree `p` is usually not a subspace of a spline space of degree `p+1` (because of the regularity constraint at the knots), so I guess this should be reformulated as :
> "
> Given the knot sequence of S a spline space of degree p, compute the knot
> sequence of S_0 a spline space of degree p+1 such that u' is in S for all u in S_0.
> Specifically, on bounded domains the first and last knots are repeated in the sequence, and in the periodic case the knot sequence is left unchanged
> "

You are correct, for S to be a subspace of S_0 we would have needed to duplicate all interior knots, which we do not do of course! I will update the docstring.

**Comment by campospinto** [comment]
> > Ok, so I guess that at some point the 'normalization' argument will be called 'basis' ?
> 
> I wanted to call it `basis`, but in this low-level module the name `basis` is used to identify an array variable that contains the values of the basis functions. So for now I decided to call it `normalization` instead, in order to avoid confusion. But as you see it is not a boolean variable any more, instead it is a string which can take the possible values `'B'` or `'M'`.
> 
> In the `SplineSpace` class in **psydac/fem/splines.py**, which is higher level, I pass the argument `basis` to the constructor. Again, this takes one of the possible values `'B'` or `'M'`.

Ok it is clear like that. And indeed, here the two bases correspond to two different normalizations

File: psydac/mapping/discrete.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 import yaml
 import yamlloader
 
-from psydac.mapping.basic import Mapping
+from sympde.topology.mapping  import Mapping
```
Comment (yguclu) [inline psydac/mapping/discrete.py:14]: Please add an empty line after this one in order to separate the external libraries from the internal library imports.

File: psydac/linalg/stencil.py [hunk 2]
```diff
@@ -337,6 +337,11 @@ def toarray( self, *, with_pads=False ):
         index = tuple( slice(p,-p) for p in self.pads )
         return self._data[index].flatten()
 
+    # ...
+    def toarray_local( self ):
+        idx = tuple( slice(p,-p) for p in self.pads )
+        return self._data[idx].flatten()
```
Comment (yguclu) [inline psydac/linalg/stencil.py:343]: This function contains exactly the same code as function `toarray` in the serial case (see 5 lines above). In order to avoid code duplication, can we have one function call the other?

Further, `toarray_local` needs a proper docstring.

File: psydac/linalg/stencil.py [hunk 3]
```diff
@@ -402,6 +407,23 @@ def _toarray_parallel_with_pads( self ):
 #        return out.flatten()
         return out.reshape(-1)
 
+    def topetsc( self ):
```
Comment (yguclu) [inline psydac/linalg/stencil.py:410]: Please add docstring.

File: psydac/linalg/stencil.py [hunk 4]
```diff
@@ -402,6 +407,23 @@ def _toarray_parallel_with_pads( self ):
...
+        space = self.space
+        assert space.parallel
```
Comment (yguclu) [inline psydac/linalg/stencil.py:412]: I think it would be very useful if the user could use the PETSc data structure and solvers also in the serial case. Can this be done?

File: psydac/linalg/stencil.py [hunk 5]
```diff
@@ -894,6 +916,59 @@ def _tocoo_no_pads( self ):
 
         return M
 
+    def tocoo_local( self ):
```
Comment (yguclu) [inline psydac/linalg/stencil.py:919]: Why do you need this function? By calling `tosparse(with_pads=False)` you will get exactly the same non-zero elements in the output sparse matrix, the only difference being that the indices will be global instead of local.

File: psydac/linalg/stencil.py [hunk 6]
```diff
@@ -980,6 +1055,38 @@ def _tocoo_parallel_with_pads( self ):
 
         return M
 
+    # ...
+    def topetsc( self ):
```
Comment (yguclu) [inline psydac/linalg/stencil.py:1059]: This function needs a docstring.

As I mentioned before, we should really handle the serial case as well.

File: psydac/linalg/kron.py [hunk 7]
```diff
@@ -102,13 +102,23 @@ def dot( self, x, out=None ):
...
     def __getitem__(self, key):
-        rows = key[:2]
-        cols = key[2:]
-        A = self._A1
-        B = self._A2
-        return A[rows[0],cols[0]]*B[rows[1],cols[1]]
+        raise NotImplementedError('TODO')
 
     # ...
     def tocoo( self ):
```
Comment (yguclu) [inline psydac/linalg/kron.py:124]: We should remove `tocoo` and `tocsr` and just add the `tosparse` method instead. In linalg we only provide one sparse format of our choice, and we want to be able to change it whenever we want. The user is free to convert the sparse matrix to any other format.

File: psydac/linalg/kron.py [hunk 8]
```diff
@@ -27,73 +29,71 @@ def kron_dot(starts, ends, pads, X, X_tmp, Y, A, B):
...
     # ...
     @property
     def codomain( self ):
-        return self._codomain
-
-    # ...
-    @property
-    def mats(self):
-        return self._mats
+        return self._space
```
Comment (yguclu) [inline psydac/linalg/kron.py:66]: This new implementation assumes that domain and codomain are the same. This looks totally wrong to me.

File: psydac/linalg/block.py [hunk 9]
```diff
@@ -201,6 +201,12 @@ def blocks( self ):
     def toarray( self ):
         return np.concatenate( [bi.toarray() for bi in self._blocks] )
 
+    def topetsc( self ):
```
Comment (yguclu) [inline psydac/linalg/block.py:204]: Please add docstring.

Can we make this work also in the serial case?

File: psydac/linalg/stencil.py (hunk 4 above)
Comment (saidctb) [reply psydac/linalg/stencil.py:412]: I assumed that everything exist in scipy in the serial case, and the user can always call it with one process in the parallel case , but probably yes it can be done,  it needs a little bit of work, because I use the information provided by the cart in ordrer for me to create the petsc cart

File: psydac/linalg/stencil.py (hunk 5 above)
Comment (saidctb) [reply psydac/linalg/stencil.py:919]: yes I need to local indices here

File: psydac/linalg/kron.py (hunk 8 above)
Comment (saidctb) [reply psydac/linalg/kron.py:66]: okay I don't remember why was this changed lol

File: psydac/api/ast/fem.py [hunk 10]
```diff
@@ -64,7 +65,6 @@
 from itertools   import groupby
```
Comment (yguclu) [inline psydac/api/ast/fem.py:65]: Please reorganize the import order as mentioned in a previous comment.

File: psydac/linalg/stencil.py (hunk 2 above)
Comment (saidctb) [reply psydac/linalg/stencil.py:343]: okay

File: psydac/api/ast/parser.py [hunk 11]
```diff
@@ -432,6 +432,7 @@ def _visit_DefNode(self, expr, **kwargs):
 
         if self.backend['name'] == 'pyccel':
             a = build_types_decorator(arguments)
+            a = [String(str(i)) for i in a]
```
Comment (yguclu) [inline psydac/api/ast/parser.py:435]: Can we merge these two lines into one? This way we avoid redefining the variable `a`.

File: psydac/api/ast/parser.py [hunk 12]
```diff
@@ -997,10 +987,10 @@ def _visit_ComputeKernelExpr(self, expr, op=None, lhs=None, **kwargs):
             ext     = target.ext if isinstance(target, Boundary) else 1
 
             J       = LogicalExpr(Jacobian(mapping), mapping=mapping, dim=dim, subs=True)
-            J       = SymbolicExpr(J)
-            values  = [ext * J.cofactor(i, j=axis) for i in range(J.shape[0])]
-            normalization = sum(v**2 for v in values)**0.5
-            values  = [v.simplify() for v in values]
+            J       = SymbolicExpr(J).inv()
```
Comment (yguclu) [inline psydac/api/ast/parser.py:990]: This variable should be called `J_inv`.

File: psydac/linalg/kron.py (hunk 7 above)
Comment (saidctb) [reply psydac/linalg/kron.py:124]: okay

File: psydac/linalg/kron.py (hunk 8 above)
Comment (saidctb) [reply psydac/linalg/kron.py:66]: in anyway, we have an assert that the domain is equal to the codomain

File: psydac/api/ast/parser.py [hunk 13]
```diff
@@ -997,10 +987,10 @@ def _visit_ComputeKernelExpr(self, expr, op=None, lhs=None, **kwargs):
...
+            values  = ext * J[axis, :]
+            normalization = values.dot(values)**0.5
+            values  = [v for v in values]
```
Comment (yguclu) [inline psydac/api/ast/parser.py:993]: It appears that the original algorithm had a bug, and it was missing the determinant factor.

But I am not convinced by these changes. The previous version was explicitly computing one row of the inverse Jacobian matrix; instead you are now computing the whole inverse and then extracting the row, which is significantly slower.

File: psydac/api/discretization.py [hunk 14]
```diff
@@ -280,132 +283,143 @@ def solve(self, **kwargs):
 
 #==============================================================================
 class DiscreteDerham(BasicDiscrete):
-    def __init__(self, *spaces):
+    """
+    Rerpresent the discrete De Rham sequence
```
Comment (yguclu) [inline psydac/api/discretization.py:287]: Typo: Rerpresent -> Represent

File: psydac/api/discretization.py [hunk 15]
```diff
@@ -280,132 +283,143 @@ def solve(self, **kwargs):
...
+            raise NotImplementedError('only global projectors are available')
+
+        if self.dim == 1:
+            P0 = Projector_H1(self.V0)
+            P1 = Projector_L2(self.V1, nquads)
+            return P0, P1
+
+        elif self.dim == 2:
+            raise NotImplementedError('TODO')
+
```
Comment (yguclu) [inline psydac/api/discretization.py:343]: It should not be a problem to provide the projectors in 2D, right?

File: psydac/api/ast/parser.py (hunk 13 above)
Comment (saidctb) [reply psydac/api/ast/parser.py:993]: yes it can be improved we actually have access to the inverse now by the new modifications in Sympde

File: psydac/api/discretization.py (hunk 15 above)
Comment (saidctb) [reply psydac/api/discretization.py:343]: yes the only problem is we need to decide which sequence the user wants in 2d

File: psydac/api/grid.py [hunk 16]
```diff
@@ -98,9 +98,10 @@ def axis(self):
     @property
     def ext(self):
         return self._ext
+
 #==============================================================================
 class BasisValues():
-    def __init__( self, V, grid, nderiv , ext=None):
+    def __init__( self, V, grid, nderiv , trial=False, ext=None):
```
Comment (yguclu) [inline psydac/api/grid.py:104]: Can you please add a docstring to this class? I think it deserves one now

File: psydac/ddm/cart.py [hunk 17]
```diff
@@ -226,6 +222,13 @@ def shape( self ):
     def subcomm( self ):
         return self._subcomm
 
+    #---------------------------------------------------------------------------
+    def topetsc( self ):
```
Comment (yguclu) [inline psydac/ddm/cart.py:226]: I know that this file does not contain many docstrings... Nevertheless, can you please add one to this method at least?

File: psydac/ddm/petsc.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+#===============================================================================
+class PetscCart:
+
+    def __init__(self, cart):
+        assert isinstance(cart, CartDecomposition)
+
+        try:
+            from petsc4py import PETSc
+        except ImportError:
+            raise
```
Comment (yguclu) [inline psydac/ddm/petsc.py:17]: Please add a docstring to this new class.

And can we raise a more intelligible error, where we say clearly that `petsc4py` must be installed with `pip` in order to use class `PetscCart`?

File: psydac/ddm/petsc.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    def create_Ao( self ):
+        cart    = self.cart
+        indices = self.indices
+        return self.petsc.AO().createBasic(indices, comm=cart.comm)
+
+    def create_LGMap( self ):
+        cart    = self.cart
+        indices = self.extended_indices
+        ao      = self.ao
+        return self.petsc.LGMap().create(ao.app2petsc(indices), comm=cart.comm)
```
Comment (yguclu) [inline psydac/ddm/petsc.py:90]: These four methods are used only by the `__init__` constructor, right? If that's correct, I think they should be "private" methods and their names should start with an underscore `_`

Comment (yguclu) [reply psydac/ddm/petsc.py:90]: A docstring would also be very useful...

File: psydac/ddm/petsc.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+    def create_g2n(self, gvec, natural):
+        cart     = self.cart
+        indices  = self.indices
+        size     = self.local_size
+        start,_  = natural.getOwnershipRange()
+
+        from_is = self.petsc.IS().createStride(size, start, 1, comm=cart.comm)
+        to_is   = self.petsc.IS().createGeneral(indices, comm=cart.comm)
+
+        return self.petsc.Scatter().create(gvec, from_is, natural, to_is)
```
Comment (yguclu) [inline psydac/ddm/petsc.py:101]: What is this method used for?

File: psydac/feec/TODO.txt [hunk 21]
```diff
@@ -0,0 +1,10 @@
+TODO:
+
+  - add core/interpolation.py (1D)
+  - add core/histopolation.py (1D)
+  - when choosing SplineSpace, choose if it should use N-basis or D-basis, remove normalize flag
+  - rename _init_collocation as _init_interpolation (also in TensorFemSpace)
+  - if N-basis, _interpolator uses collocation matrix
+  - if D-basis, _interpolator uses histopolation matrix
+
+  - remove normalize argument from discretize_space (and everywhere else)
```
Comment (yguclu) [inline psydac/feec/TODO.txt:10]: Can we please check if we have done everything in this TODO list? If so, I think we can remove this file altogether.

File: psydac/ddm/petsc.py (hunk 18 above)
Comment (saidctb) [reply psydac/ddm/petsc.py:17]: it's more difficult than that, for it to work  the user needs to install petsc first and then petsc4py , I think the import error should be enough

File: psydac/feec/global_projectors.py [hunk 22]
```diff
@@ -0,0 +1,472 @@
...
+
+        # build the rhs
+        self.func(*self.args, fun)
+
+        if len(self.N)==1:
+            rhs = self.rhs.toarray()
+            coeffs = array_to_stencil(self.N[0].solve(rhs), self.space.vector_space)
+            coeffs.update_ghost_regions()
+        else:
+            coeffs = kronecker_solve(solvers = self.N, rhs = self.rhs)
```
Comment (yguclu) [inline psydac/feec/global_projectors.py:64]: As you can see [here](https://github.com/pyccel/psydac/blob/fb5f9d6552ced9ccc3d552f2fa6f8096bd70d6a8/psydac/linalg/kron.py#L417) this function accepts an `out` keyword argument. Hence this line should read
```python
kronecker_solve(solvers = self.N, rhs = self.rhs, out=coeffs)
```

File: psydac/feec/global_projectors.py [hunk 23]
```diff
@@ -0,0 +1,472 @@
...
+
+        # build the rhs
+        self.func(*self.args, *fun)
+
+        self.rhs.update_ghost_regions()
+
+        coeffs    = BlockVector(self.space.vector_space)
+        coeffs[0] = kronecker_solve(solvers = self.DNN, rhs = self.rhs[0])
+        coeffs[1] = kronecker_solve(solvers = self.NDN, rhs = self.rhs[1])
+        coeffs[2] = kronecker_solve(solvers = self.NND, rhs = self.rhs[2])
```
Comment (yguclu) [inline psydac/feec/global_projectors.py:129]: Same comment as above, `out=coeffs[i]` with `i=0,1,2` should be passed as an argument.

File: psydac/feec/global_projectors.py [hunk 24]
```diff
@@ -0,0 +1,472 @@
...
+
+        # build the rhs
+        self.func(*self.args, *fun)
+
+        self.rhs.update_ghost_regions()
+
+        coeffs    = BlockVector(self.space.vector_space)
+        coeffs[0] = kronecker_solve(solvers = self.NDD, rhs = self.rhs[0])
+        coeffs[1] = kronecker_solve(solvers = self.DND, rhs = self.rhs[1])
+        coeffs[2] = kronecker_solve(solvers = self.DDN, rhs = self.rhs[2])
```
Comment (yguclu) [inline psydac/feec/global_projectors.py:196]: Again, you can pass `out=coeffs[i]`

File: psydac/feec/global_projectors.py [hunk 25]
```diff
@@ -0,0 +1,472 @@
...
+
+        # build the rhs
+        self.func(*self.args, fun)
+
+        if len(self.D) == 1:
+            rhs = self.rhs.toarray()
+            coeffs = array_to_stencil(self.D[0].solve(rhs), self.space.vector_space)
+            coeffs.update_ghost_regions()
+        else:
+            coeffs = kronecker_solve(solvers = self.D, rhs = self.rhs)
```
Comment (yguclu) [inline psydac/feec/global_projectors.py:258]: Again, use `out=coeffs`

File: psydac/fem/splines.py [hunk 26]
```diff
@@ -11,15 +11,37 @@
...
+        xleft  = xgrid[:-1]
+        xright = xgrid[1:]
+    elif xgrid[0] != xA:
+        xleft  = np.array([xgrid[-1] - (xB-xA), *xgrid[:-1]])
+        xright = xgrid
+    elif xgrid[-1] != xB:
+        xleft  = xgrid
+        xright = np.array([*xgrid[1:], xgrid[0] + (xB-xA)])
+
+    return xleft, xright
```
Comment (yguclu) [inline psydac/fem/splines.py:43]: Can you please check if this function is ever used in Psydac? (I think it's a leftover from a previous version of this file)

If it's not used, feel free to remove it.

File: psydac/ddm/petsc.py (hunk 20 above)
Comment (saidctb) [reply psydac/ddm/petsc.py:101]: go from global ordering  to natural ordering

File: psydac/fem/splines.py (hunk 26 above)
Comment (saidctb) [reply psydac/fem/splines.py:43]: okay

File: psydac/fem/tensor.py [hunk 27]
```diff
@@ -418,52 +435,56 @@ def compute_interpolant( self, values, field ):
 
     def reduce_grid(self, axes=[], knots=[]):
```
Comment (yguclu) [inline psydac/fem/tensor.py:436]: Default argument values must be immutable. Please use empty tuples `()`, or set them to `None`.

File: psydac/feec/tests/todo_test_derivatives.py [hunk 28]
```diff
@@ -1,49 +1,105 @@
-from psydac.feec.derivatives import Grad, Curl, Div
-from psydac.fem.tensor       import TensorFemSpace
-from psydac.fem.splines      import SplineSpace
-from psydac.linalg.stencil   import StencilVector
-from psydac.linalg.block     import BlockVector,ProductSpace
-from psydac.fem.vector       import ProductFemSpace
-from psydac.feec.utilities   import Interpolation, interpolation_matrices
+import pytest
```
Comment (yguclu) [inline psydac/feec/tests/todo_test_derivatives.py:1]: I think that this file can be removed.

File: psydac/feec/tests/test_commuting_projections.py [hunk 29]
```diff
@@ -0,0 +1,209 @@
+# -*- coding: UTF-8 -*-
```
Comment (yguclu) [inline psydac/feec/tests/test_commuting_projections.py:1]: I think that we should also add tests in 1D and 2D.

File: psydac/feec/tests/OLD_test_differentiation_matrices.py [hunk 30]
```diff
@@ -0,0 +1,417 @@
+import numpy as np
```
Comment (yguclu) [inline psydac/feec/tests/OLD_test_differentiation_matrices.py:1]: We should remove this old file.

File: psydac/feec/OLD_derivatives.py [hunk 31]
```diff
@@ -0,0 +1,318 @@
+# -*- coding: UTF-8 -*-
```
Comment (yguclu) [inline psydac/feec/OLD_derivatives.py:1]: We should remove this old file.

File: psydac/feec/struphy_example.py [hunk 32]
```diff
@@ -0,0 +1,70 @@
+from mpi4py import MPI
```
Comment (yguclu) [inline psydac/feec/struphy_example.py:1]: I think we have a more complete example in our unit tests, right? So we can remove this old file.

File: psydac/feec/pull_push.py [hunk 33]
```diff
@@ -0,0 +1,215 @@
+from sympy import lambdify
```
Comment (yguclu) [inline psydac/feec/pull_push.py:1]: Every function in this file only works for 3D mappings. Maybe we should call it `push_pull_3d.py`?

We definitely need 1D and 2D versions as well.

Further, there are no docstrings...

File: psydac/api/tests/test_api_feec_3d.py [hunk 34]
```diff
@@ -0,0 +1,206 @@
...
+    return error
+
+###############################################################################
+#            SERIAL TESTS
+###############################################################################
+
+#==============================================================================
+# 3D Maxwell's equations with "Collela" map
+#==============================================================================
+def test_maxwell_3d_1():
```
Comment (yguclu) [inline psydac/api/tests/test_api_feec_3d.py:156]: We need to add a non-periodic test. A standing wave in a cavity would work nicely...

File: psydac/linalg/stencil.py (hunk 5 above)
Comment (yguclu) [reply psydac/linalg/stencil.py:919]: I see. This function is almost identical to `tosparse`, and it would be nice to avoid such a large code duplication.

Is there not a way to have the `tosparse` function return what you need, maybe by passing the flag `local=True`?

File: psydac/linalg/kron.py (hunk 8 above)
Comment (yguclu) [reply psydac/linalg/kron.py:66]: I think that assert is new, it should not be there

File: psydac/api/ast/parser.py (hunk 13 above)
Comment (yguclu) [reply psydac/api/ast/parser.py:993]: OK, let's fix it in this PR.

File: psydac/api/discretization.py (hunk 15 above)
Comment (yguclu) [reply psydac/api/discretization.py:343]: Sure, but the user tells us explicitly which 2D DeRham sequence is needed when the symbolic DeRham sequence is created.

File: psydac/ddm/petsc.py (hunk 18 above)
Comment (yguclu) [reply psydac/ddm/petsc.py:17]: But why are you using a try/except clause, if not for giving a nice error for the user??

File: psydac/ddm/petsc.py (hunk 20 above)
Comment (yguclu) [reply psydac/ddm/petsc.py:101]: I don't think that this method is ever used in Psydac... Can you check?

File: psydac/api/tests/test_api_feec_3d.py (hunk 34 above)
Comment (saidctb) [reply psydac/api/tests/test_api_feec_3d.py:156]: this will need more work in order to apply boundary conditions and it's not something we currently have with the scipy format

File: psydac/feec/pull_push.py (hunk 33 above)
Comment (saidctb) [reply psydac/feec/pull_push.py:1]: yes we only have the 3d

File: psydac/feec/global_projectors.py (hunk 23 above)
Comment (saidctb) [reply psydac/feec/global_projectors.py:129]: the coeffs[i] is not allocated so it's not gonna work

File: psydac/feec/pull_push.py (hunk 33 above)
Comment (saidctb) [reply psydac/feec/pull_push.py:1]: we should keep the same name, and add the other cases later

File: psydac/linalg/kron.py (hunk 8 above)
Comment (saidctb) [reply psydac/linalg/kron.py]: okay this class is for 2d stencil matrix, it's not used at all, so it doen't really matter, we are using the general one, we just need to keep this one for the moment until we improve the general, and then we can delet it

File: psydac/api/discretization.py (hunk 15 above)
Comment (saidctb) [reply psydac/api/discretization.py:341]: yes that's true, we can add it

File: psydac/ddm/petsc.py (hunk 19 above)
Comment (saidctb) [reply psydac/ddm/petsc.py]: done

File: psydac/linalg/kron.py (hunk 8 above)
Comment (yguclu) [reply psydac/linalg/kron.py]: OK. Please comment it out so that it's clear that we are not using it. Or maybe put it in a separate file to be removed later.

Comment (saidctb) [reply psydac/linalg/kron.py]: I will comment it out and when we don't need it, we can delete it

File: psydac/api/grid.py (hunk 16 above)
Comment (yguclu) [reply psydac/api/grid.py:104]: Thanks for adding the docstring. But the constructor's docstring should be placed outside of the `__init__` function. In other words, this should be the class' docstring.

File: psydac/ddm/petsc.py (hunk 18 above)
Comment (yguclu) [reply psydac/ddm/petsc.py:17]: If you use a `try/except` block, you should give a proper error. Otherwise you can just remove the `try/except` block...

Comment (saidctb) [reply psydac/ddm/petsc.py:17]: yes that's true

File: psydac/linalg/stencil.py (hunk 4 above)
Comment (yguclu) [reply psydac/linalg/stencil.py:415]: Please create a separate issue: "make petsc backend work in serial".

File: psydac/ddm/petsc.py (hunk 18 above)
Comment (saidctb) [reply psydac/ddm/petsc.py]: done

File: psydac/api/grid.py (hunk 16 above)
Comment (saidctb) [reply psydac/api/grid.py]: done

File: psydac/feec/tests/test_commuting_projections.py (hunk 29 above)
Comment (saidctb) [reply psydac/feec/tests/test_commuting_projections.py:1]: done

**Review by saidctb — DISMISSED** [review]

File: psydac/linalg/stencil.py (hunk 4 above)
Comment (yguclu) [reply psydac/linalg/stencil.py:415]: Thanks for creating #79

File: psydac/ddm/petsc.py [hunk 35]
```diff
@@ -0,0 +1,112 @@
...
+
+import numpy as np
+from itertools import product
+
+from .cart import CartDecomposition
+
+#===============================================================================
+class PetscCart:
+
+    def __init__(self, cart):
```
Comment (yguclu) [inline psydac/ddm/petsc.py:11]: Can you please add a docstring to this nice class?

As usual this should contain the `Parameters` section, and a list of the properties under the `Attributes` section (https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring).

File: psydac/api/tests/test_api_feec_3d.py (hunk 34 above)
Comment (yguclu) [reply psydac/api/tests/test_api_feec_3d.py:261]: OK. Let's not worry about it in this pull request.

**Review by saidctb — APPROVED** [review]

File: psydac/ddm/petsc.py (hunk 35 above)
Comment (yguclu) [reply psydac/ddm/petsc.py:11]: Let's leave this to issue #79. Things should be properly documented there.
